[decisions] headers (#4246)
ElektraInitiative/libelektra: doc/decisions/header_file_structure.md

      Each of these functions/structs/symbols has a specific "target audience" that needs access.

      Symbols belonging to this category should not appear in a public section of the `symbols.map` file.

  4.  Finally, there things that aren't part of the public API, but may be in future.
      This includes e.g., `ksFindHierarchy` or `elektraReadArrayNumber`.
      These functions could be public, but for various reasons are not.
      Maybe they are not well-tested, or maybe we just don't want to commit to the function yet.

      Symbols belonging to this category should not appear in a public section of the `symbols.map` file.

## Considered Alternatives

## Decision

### Libraries

A library can be monolithic or modularized.
Monolithic libraries should be small and bigger libraries with large APIs should be modularized.

A monolithic library `foo` may have these headers (covering categories 3 & 4 from above):

- `src/include/elektra/foo.h`:
  Contains the full public API of `libelektra-foo`.
  Will be installed as `<include-root>/elektra/foo.h`.
- `src/include/elektra/foo/internal.h`:
  Contains the internal (i.e., non-stable) API of `libelektra-foo`.

>   Will be installed as `<include-root>/elektra/foo/internal.h`.

Reviewer: I don't think we should install this file.

> A modularized library `bar` may have these headers (covering categories 3 & 4 from above):

Reviewer: ```suggestion
A modularized library `foo` may have these headers (covering categories 3 & 4 from above):
```

> - `src/include/elektra/foo/public.h`:

Reviewer: I don't like the word public here. What about  "foo/foo.h" to include everything?

Author: I'm not a massive fan of the repetition, but I guess it's better than `public`.

Reviewer: I also considered "all" and "everything".

  Declares the public API of `libelektra-foo`, by including `#include <elektra/foo/*.h>`.

>   Will be installed as `<include-root>/elektra/public/foo.h`.

Reviewer: This sounds weird to me. I wouldn't restructure the files during installation.

Author: That's actually a typo. It was supposed to be

```suggestion
  Will be installed as `<include-root>/elektra/foo/public.h`.
```

  Such a header may only include `#include <elektra/foo/*.h>` lines.

- `src/include/elektra/foo/*.h`:
  Additional public API header of `libelektra-foo`.
  Will be installed as `<include-root>/elektra/foo/*.h`.

  All of these installed headers `installed.h` must be included from `src/lib/foo/public.h` via a line `#include <elektra/foo/installed.h>`.

- `src/include/elektra/foo/internal.h`:
  Declares the internal (i.e., non-stable) API of `libelektra-foo`, by including `#include <elektra/foo/*.h>`.

>   Will be installed as `<include-root>/elektra/internal/foo.h`.

Reviewer: Again, it is okay to have internal header files (to be used within other modules of Elektra) but they should not be installed.

  Such a header may only include `#include <elektra/foo/*.h>` lines.

- `src/include/elektra/foo/internal/*.h`:
  Additional header of `libelektra-foo`.
  Will be installed as `<include-root>/elektra/foo/internal/*.h`.

  All of these installed headers `installed.h` must be included from `src/lib/foo/internal.h` via a line `#include <elektra/foo/internal/installed.h>`.

Additionally, all libraries may also have private headers:

- `src/lib/foo/**/*.h`:
  Additional headers may be present.
  These headers may only be used by other files within `src/lib/foo`, according to the rules in [Including Headers](header_include.md).

### Plugins

Plugins do not declare their API via header files.

> Their headers are never installed and can be named any way the developer wants.

Reviewer: :+1:

> ### Tests

Reviewer: This info also needs to be in doc/TESTING.md

For category 2 from above (private but needs to be tested), one of the following should be done:

1. Declare functions as `static` in a `.c` file and `#include ""` this file from the test.
2. Add a private non-installed header (e.g. `src/lib/foo/testapi.h`) that declares the API that needs testing, `#include ""` that and compile the test sources together with the `.o` files from the library (static linking).

If a symbol is needed in only one file and for tests, option 1 should be preferred.
For symbols that are used in multiple files, a header needs to exist anyway.
In any case, these unit tests should not be installed and should statically link the library into the test executable.
This way we don't pollute our `symbols.map` files and keep the number of exported symbols down.

## Rationale

- This structure makes the `#include`s simple and works nicely with our directory structure.
- Having a uniform naming convention simplifies things, both for developers writing Elektra and those using Elektra.

> - Requiring libraries must either be fully modularized (main headers are only `#includes`) or completely monolithic, avoids the situation where a library has some parts in `public.h` and other parts in extra headers.

Reviewer: :+1:

## Constraints

- Each library should have at least one separate header file:

  Without this constraint, we would constantly include things that aren't needed.
  This is inconvenient as it clutters the auto-suggestions of IDEs.

>   It also slows down the compile process, since the compiler has to parse everything that was included (see [also](https://lore.kernel.org/lkml/YdIfz+LMewetSaEB@gmail.com/T/)).

Reviewer: The link is misleading here. This was a huge undertaking and has little to do what is planned with this decision. Please move it down to Notes.

- Non-public API and public API should be kept in separate files
- Some non-public APIs need to be accessible for testing, such headers should not be packaged/installed

## Assumptions

- There are different categories of "private":

  1.  Some parts are truly private, i.e., shouldn't be used outside the library that defines them.
      These things are only there because a library was split into multiple `.c` files.

>       This includes e.g., `splitNew` and `struct _Split`.

Reviewer: Do they still exist?

Author: No, but `backendsMerge` and `struct _BackendData` do, I'll change that.

  3.  Some things are not part of the public API and will never be part of the public API.
      This includes the `struct _Key` and `struct _KeySet`, but `elektraMalloc` and the high-level functions needed for codegen.
      These things will never be public API, because we don't want to make all the guarantees associated with that.
      Nonetheless, they cannot truly be private, because functions in other libraries need access.
      Each of these functions/structs/symbols has a specific "target audience" that needs access.

      Symbols belonging to this category should not appear in a public section of the `symbols.map` file.

  4.  Finally, there things that aren't part of the public API, but may be in future.
      This includes e.g., `ksFindHierarchy` or `elektraReadArrayNumber`.
      These functions could be public, but for various reasons are not.
      Maybe they are not well-tested, or maybe we just don't want to commit to the function yet.

      Symbols belonging to this category should not appear in a public section of the `symbols.map` file.

## Considered Alternatives

## Decision

> ### Libraries

Reviewer: The title as-is is confusing.

```suggestion
### Header Files for Different Types of Libraries
```

Author: I find your title a bit verbose and it would also no longer match the next sections "Plugins" and "Tools".

How about, I add an explanation above e.g.

```suggestion
The decision is split into 4 subsections "Libraries", "Plugins", "Tools" and "Tests", because the different parts of Elektra have different requirements.

### Libraries
```

Note: I'll also add the section about tools. It will basically say, something similar to the subsection about plugins. Tools don't have a public API (at least not a C API), so no restrictions apply.

Reviewer: Yes, makes it much clearer! :+1:

A library can be monolithic or modularized.
Monolithic libraries should be small and bigger libraries with large APIs should be modularized.

A monolithic library `foo` may have these headers (covering categories 3 & 4 from above):

- `src/include/elektra/foo.h`:
  Contains the full public API of `libelektra-foo`.
  Will be installed as `<include-root>/elektra/foo.h`.
- `src/include/internal/foo.h`:
  Contains the internal API of `libelektra-foo`.
  Will not be installed.

A modularized library `foo` may have these headers (covering categories 3 & 4 from above):

- `src/include/elektra/foo.h`:
  Declares the public API of `libelektra-foo`, by including `#include <elektra/foo/*.h>`.
  Will be installed as `<include-root>/elektra/foo/public.h`.

  Such a header may only contain `#include <elektra/foo/*.h>` lines.

- `src/include/elektra/foo/*.h`:
  Additional public API header of `libelektra-foo`.
  Will be installed as `<include-root>/elektra/foo/*.h`.

>   All of these installed headers `installed.h` must be included from `src/include/elektra/foo.h` via a line `#include <elektra/foo/installed.h>`.

Reviewer: Sentence is broken

Author: I don't see the problem...

Would this make it better:

```suggestion
  All of these "installed headers" (`installed.h`) must be included from `src/include/elektra/foo.h` via a line `#include <elektra/foo/installed.h>`.
```

Reviewer: It is a bit better but still: "installed headers" is plural, `installed.h` is one header. Maybe simply a "for example" is missing to fix the sentence but I am not sure what you want to say here at the moment.

Author: How about 

```suggestion
  Any one `installed.h` of these installed headers must be included from `src/include/elektra/foo.h` via a line `#include <elektra/foo/installed.h>`.
```

Reviewer: Yes, better.

- `src/include/internal/foo.h`:
  Declares the internal API of `libelektra-foo`, by including `#include <internal/foo/*.h>`.
  Will not be installed.

  Such a header may only contain `#include <internal/foo/*.h>` lines.

- `src/include/internal/foo/*.h`:
  Additional internal header of `libelektra-foo`.
  Will not be installed.

>   All of these installed headers `installed.h` must be included from `src/include/internal/foo.h` via a line `#include <internal/foo/installed.h>`.

Reviewer: Also here.

If a symbol is needed in only one file and for tests, option 1 should be preferred.
For symbols that are used in multiple files, a header needs to exist anyway.
In any case, these unit tests should not be installed and should statically link the library into the test executable.
This way we don't pollute our `symbols.map` files and keep the number of exported symbols down.

## Rationale

- This structure makes the `#include`s simple and works nicely with our directory structure.
- Having a uniform naming convention simplifies things, both for developers writing Elektra and those using Elektra.
- Requiring libraries must either be fully modularized (main headers are only `#includes`) or completely monolithic, avoids the situation where a library has some parts in `foo.h` and other parts in extra headers.
  This is bad, because it encourages users to simply include `foo.h`, which defeats the point of modularized headers.

## Implications

- The location of a header file within the source tree determines what API it contains:
  - Public: Will be installed, must be stable, can be used anywhere

>   - Internal: Will not be installed, doesn't have to be stable, can be used anywhere, but public headers

Reviewer: ```suggestion
  - Internal: Will not be installed, doesn't have to be stable, can be used anywhere but public headers
```

>   - Private: Will not be installed, doesn't have to be stable, can only be used in the same library

Reviewer: This private is different to the 4 private above?

Author: It's category 1. Truly private shouldn't be used outside library.

## Constraints

- Each library should have at least one separate header file:

  Without this constraint, we would constantly include things that aren't needed.
  This is inconvenient as it clutters the auto-suggestions of IDEs.
  It also slows down the compile process, since the compiler has to parse everything that was included.

- Non-public API and public API should be kept in separate files

> - Some non-public APIs need to be accessible for testing, such headers should not be packaged/installed

Reviewer 2: What a coincidence! I just stumbled upon this while creating the high-level change-tracking API. Currently I just added a function for creating a fake/mock ChangeTrackingContext into the public API, but I thought maybe we should have seperate `-testing` headers for such functionality? This then would have to be installed, as 3rdpartys may write plugins that use it.

Author: > This then would have to be installed, as 3rdpartys may write plugins that use it.

Well, if the API is intended for consumption by third parties, then it should be installed anyway and not a separate header for testing. If people need more, they should just copy the headers from the repo into their code.

## Assumptions

- There are different categories of "private":

  1.  Some parts are truly private, i.e., shouldn't be used outside the library that defines them.
      These things are only there because a library was split into multiple `.c` files.
      This includes e.g., `opmphmNew` or `struct _BackendData`.

      Symbols belonging to this category should not appear at all in the `symbols.map` file.
      Headers declaring such symbols must never be installed.

  2.  Other things are truly private, but must be tested.
      This includes e.g., most `backends*` functions or the `elektraKeyName*` functions.

      Symbols belonging to this category should not appear in at all in the `symbols.map` file.
      Headers declaring such symbols must never be installed.

>   3.  Some things are not part of the public API and will never be part of the public API.

Reviewer 2: So this is visibility 'internal'?

## Constraints

- Each library should have at least one separate header file:

  Without this constraint, we would constantly include things that aren't needed.
  This is inconvenient as it clutters the auto-suggestions of IDEs.
  It also slows down the compile process, since the compiler has to parse everything that was included (see [also](https://lore.kernel.org/lkml/YdIfz+LMewetSaEB@gmail.com/T/)).

> - Non-public API and public API should be kept in separate files

Author: Non-public in this case includes things like the definitions of `struct _Key` et al. So in some cases it makes sense to install the non-public headers. In other cases (like the `fortests`) we probably don't need to install them.

- Some non-public APIs need to be accessible for testing, such headers should not be packaged/installed

## Assumptions

> - There are different categories of "private":

Reviewer: This is a very important point, actually it is a decision on its own.

Author: What exactly would the decision part of the decision be? I'm just stating facts here. The assumption is that all symbols can be put into one of these four groups. But I'm not sure there is anything to decide here. The decision to use these four groups to build our headers is what **this** decision is about. Not sure, what you want to separate here.

Reviewer: It is probably more standard to use the word "protected" for something that is "public" within Elektra's repository but private for the outside world (point 3).

> The assumption is that all symbols can be put into one of these four groups.

Then write what you mean:

```suggestion
- We assume that all parts of the API belong to one of the following visibility categories:
```

> Not sure, what you want to separate here.

The separate decision would be about which parts belong to which "visibility category".  You give good examples here.

Reviewer 2: Hmm honestly I don't think we should do a seperate decision here. The examples that @kodebach listed serve illustrative purposes in this decision. 

Creating decisions just for documenting where which each function of each library should be placed is too much IMHO. That sounds very much like documentation for documentation's sake. Checking the placement is something that should be part of the PR review.

Reviewer 2: If we apply the suggestion by @markus2330 then we must also define 'public' and 'internal' visibilities in this section.

  1.  Some parts are truly private, i.e. shouldn't be used outside the library that defines them.
      These things are only there because a library was split into multiple `.c` files.
      This includes e.g. `splitNew` and `struct _Split`.

      Symbols belonging to this category should not appear at all in the `symbols.map` file.

  2.  Other things are truly private, but must be tested.
      This includes e.g. most other `split*` functions or the `elektraKeyName*` functions.

>       Symbols belonging to this category should not appear in a public section of the `symbols.map` file.

Reviewer: Do we need these symbols at all? In theory unit tests could simply include the .c files (with static functions).

Author: IMO including .c files is a bad idea.

1. It means that there will be a separate definition of the function in every file that includes the function. The performance aspect wouldn't matter for tests, but it would definitely break things like `static` variables (if we have those anywhere).
2. I think it's harder to define rules for when including `.c` files is okay and which files can be included where, than to just say "if you need to test a private function, make it non-static, add it to a `fortests` header and add it to the `fortests` section of the `symbols.map`".
